### PR TITLE
Emit Eval progress in console logging

### PR DESCRIPTION
**Pull request by eracah**
# What does this PR do?

* Adds a callback event for `eval_batch_end` that prints the cur_eval_batch/total_eval_batches progress. 
* Logging frequency is based on the deciles are the total number of eval batches.
* Along for the ride: changes default `console_log_level` to `1ba`

# Testing
* Added unit test
* did manual local test
* manual test on cluster with gpt-125m from mosaicml llm examples-repo:
```python
composer main.py yamls/mosaic_gpt/125m.yaml \
    train_loader.dataset.split=train_small \
    global_train_batch_size=32 \
    device_train_microbatch_size=1 \
    fsdp_config.activation_checkpointing=true \
    fsdp_config.activation_cpu_offload=true \
    max_duration=10ba \
    callbacks.speed_monitor.window_size=1 \
    eval_interval=2ba \
    eval_subset_num_batches=320 \
    save_folder='./checkpoints' \
    save_interval=5ba \
    loggers='{wandb:{}}' \
    run_name=$COMPOSER_RUN_NAME
```
results below:

```bash
[batch=1/10]:
	 Train epoch: 0
	 Train trainer/global_step: 0
	 Train trainer/batch_idx: 0
          .
          .
          .
[batch=2/10]:
	 Train trainer/global_step: 1
	 Train trainer/batch_idx: 1
          .
          .
          .
[Eval batch=1/320] Eval on eval data
[Eval batch=33/320] Eval on eval data
[Eval batch=65/320] Eval on eval data
[Eval batch=97/320] Eval on eval data
[Eval batch=129/320] Eval on eval data
[Eval batch=160/320] Eval on eval data
[Eval batch=192/320] Eval on eval data
[Eval batch=224/320] Eval on eval data
[Eval batch=256/320] Eval on eval data
[Eval batch=288/320] Eval on eval data
[Eval batch=320/320] Eval on eval data
[Eval batch=320/320] Eval on eval data:
	 Eval LanguageCrossEntropy: 10.9731
	 Eval Perplexity: 58297.7109
[batch=3/10]:
	 Train epoch: 0
	 Train trainer/global_step: 2
```

# What issue(s) does this change relate to?
fix CO-1615
fix CO-1704

### CODE REVIEW

**Comment by eracah** [comment]
> I think this LGTM, but tbh I don't really understand console logger.... Assuming u manually tested as well

Yes. See PR description

File: composer/loggers/console_logger.py [hunk 1]
```diff
@@ -57,6 +57,7 @@ def __init__(self,
             else:
                 raise ValueError(f'stream must be one of ("stdout", "stderr", TextIO-like), got {stream}')
         self.should_log_traces = log_traces
+        self.eval_log_interval = Time.from_timestring('100ba')
```
Comment (dakinggg) [inline composer/loggers/console_logger.py:60]: I'm thinking this should probably not be hardcoded?

File: composer/loggers/console_logger.py [hunk 2]
```diff
@@ -144,10 +160,13 @@ def _get_progress_string(self, state: State):
...
-    def log_to_console(self, data: Dict[str, Any], state: State, prefix: str = '') -> None:
+    def log_to_console(self, data: Dict[str, Any], state: State, prefix: str = '', is_train=True) -> None:
         # log to console
-        training_progress = self._get_progress_string(state)
-        log_str = f'{training_progress}:'
+        if is_train:
+            progress = self._get_progress_string(state)
+        else:
+            progress = self._get_eval_progress_string(state)
+        log_str = f'{progress}:'
```
Comment (dakinggg) [inline composer/loggers/console_logger.py:169]: Can we not emit the `:` if there is nothing coming after that? It looks a bit strange in the logs.

Comment (eracah) [reply composer/loggers/console_logger.py]: fixed

File: composer/loggers/console_logger.py (hunk 1 above)
Comment (eracah) [reply composer/loggers/console_logger.py]: moved to deciles.

File: composer/loggers/console_logger.py [hunk 3]
```diff
@@ -117,10 +127,32 @@ def predict_start(self, state: State, logger: Logger) -> None:
...
+        eval_dataloader_label = state.dataloader_label
+        total_eval_batches = self._get_total_eval_batches(state)
+        curr_progress = f'[Eval batch={eval_batch}/{total_eval_batches}] Eval on {eval_dataloader_label} data'
+        return curr_progress
+
+    def _get_total_eval_batches(self, state: State) -> int:
+        cur_evaluator = [evaluator for evaluator in state.evaluators if evaluator.label == state.dataloader_label][0]
+        total_eval_batches = int(
+            state.dataloader_len) if state.dataloader_len is not None else cur_evaluator.subset_num_batches
+        # To please pyright. Based _set_evaluator_interval_and_subset_num_batches, total_eval_batches can't be None
```
Comment (dakinggg) [inline composer/loggers/console_logger.py:152]:
```suggestion
        # To please pyright. Based on _set_evaluator_interval_and_subset_num_batches, total_eval_batches can't be None
```

**Review by dakinggg — APPROVED** [review]
LGTM

File: composer/loggers/console_logger.py [hunk 4]
```diff
@@ -117,10 +127,32 @@ def predict_start(self, state: State, logger: Logger) -> None:
             self._log_hparams_to_console()
 
     def eval_start(self, state: State, logger: Logger) -> None:
+        total_eval_batches = self._get_total_eval_batches(state)
+        deciles = np.linspace(0, 1, NUM_EVAL_LOGGING_EVENTS + 1)
+        batch_idxs = np.arange(1, total_eval_batches + 1)
+        if total_eval_batches < NUM_EVAL_LOGGING_EVENTS:
+            self.eval_batch_idxs_to_log = list(batch_idxs)
+        else:
+            self.eval_batch_idxs_to_log = list(np.quantile(batch_idxs, deciles).round().astype(dtype=int))
```
Comment (dakinggg) [inline composer/loggers/console_logger.py:136]: oh just realized one thing. This should be a set I think? Otherwise you duplicate the last print line

Comment (eracah) [reply composer/loggers/console_logger.py:136]: because of the eval_end event?

Comment (dakinggg) [reply composer/loggers/console_logger.py:136]: yeah, eg in your example logs
```
[Eval batch=320/320] Eval on eval data
[Eval batch=320/320] Eval on eval data:
	 Eval LanguageCrossEntropy: 10.9731
	 Eval Perplexity: 58297.7109
	 ```

Comment (eracah) [reply composer/loggers/console_logger.py:136]: oh right I think maybe just `eval_batch_idxs.remove(num_eval_batches)` perhaos?

Comment (dakinggg) [reply composer/loggers/console_logger.py:136]: why not make it a set? It is logically a set, and also the 'in' operation will be (ever so slightly) faster

Comment (eracah) [reply composer/loggers/console_logger.py:136]: but the issue isn't that `self.eval_batch_idxs_to_log` has duplicate values. It's that `eval_batch_end` runs at the end of the last batch and then `eval_end` runs right after

Comment (eracah) [reply composer/loggers/console_logger.py:136]: so if we remove the last element of the list then `eval_batch_end` won't run at the end of the last batch

Comment (eracah) [reply composer/loggers/console_logger.py:136]: ok thanks good catch fixed

**Pull request merged**